src/main/java/org/spdx/utility/compare/TemplateRegexGenerator.java

```diff
@@ -0,0 +1,322 @@
+	
+	static final Logger logger = LoggerFactory.getLogger(TemplateRegexGenerator.class);
+	
+	static final int WORD_LIMIT = 25; // number of words to search for at the beginning and end of the template
+	
+	static abstract class RegexElement {
+	}
+	
+	static class RegexList extends RegexElement {
+		private List<RegexElement> elements = new ArrayList<>();
+		
+		public void addElement(RegexElement element) {
+			elements.add(element);
+		}
+		
+		public List<RegexElement> getElements() {
+			return elements;
+		}
+		
+		@Override
+		public String toString() {
+			StringBuilder sb = new StringBuilder();
+			for (RegexElement element:elements) {
+				sb.append(element.toString());
+			}
+			return sb.toString();
+		}
+	}
+	
+	/**
+	 * Group with the optional modifier
+	 */
+	static class OptionalRegexGroup extends RegexList {
+		
+		@Override
+		public String toString() {
+			return "(" + super.toString() + ")?";
+		}
+	}
+	
+	static class RegexToken extends RegexElement {
+		private String token;
+		
+		public RegexToken(String token) {
+			this.token = token.trim();
+		}
+		
+		public String getToken() {
+			return token;
+		}
+		
+		@Override
+		public String toString() {
+			return Pattern.quote(
+					LicenseCompareHelper.NORMALIZE_TOKENS.getOrDefault(token.toLowerCase(), token)) + 
+					"\\s*";
+		}
+	}
+	
+	static class RegexPattern extends RegexElement {
+		private String pattern;
+		
+		public RegexPattern(String pattern) {
+			this.pattern = pattern;
+		}
+		
+		public String getPattern() {
+			return pattern;
+		}
+		
+		@Override
+		public String toString() {
+			return "(" + pattern + ")"; // We always treat the pattern as a group
+		}
+
+		/**
+		 * @param pattern pattern to set
+		 */
+		public void setPattern(String pattern) {
+			this.pattern = pattern;
+		}
+	}
+
+	private String template;
+	
+	/**
+	 * Top level regex
+	 */
+	private RegexList regexPatternList = new RegexList();
+	
+	
+	private int optionalNestLevel = 0;
+	
+	private List<OptionalRegexGroup> optionalGroups = new ArrayList<>();
+
+	/**
+	 * Generates regular expressions from a license or exception template
+	 * @throws SpdxCompareException 
+	 */
+	public TemplateRegexGenerator(String template) throws SpdxCompareException {
+		this.template = template;
+		parseTemplate();
+	}
+	
+	/**
+	 * Parses the template generating the regular expression
+	 * @throws SpdxCompareException 
+	 */
+	private void parseTemplate() throws SpdxCompareException {
+		try {
+			SpdxLicenseTemplateHelper.parseTemplate(LicenseCompareHelper.removeCommentChars(template), this);
+		} catch (LicenseTemplateRuleException e) {
+			throw new SpdxCompareException("Invalid template rule found during filter: "+e.getMessage(),e);
+		} catch (LicenseParserException e) {
+			throw new SpdxCompareException("Invalid template found during filter: "+e.getMessage(),e);
+		}
+	}
+	
+	/**
+	 * @return the complete regular expression for the template
+	 */
+	public String getCompleteRegex() {
```

Comment:
I wonder if there should be a convenience wrapper method provided for this method, which compiles the result of `getCompleteRegex` into a `Pattern`?  That way we can ensure that any flags / modes that are needed (e.g. `DOTALL` mode) are set in one place (i.e. inside `TemplateRegexGenerator`).

That said, I still think returning a `String` representation of the regex is also useful and should be part of the public API, but _also_ providing a `Pattern` version as well seems to me to be valuable (e.g. it may also make it easier to switch from programmatically setting modes to using the inline mode flags inside the regex String itself, if we decided to do that).

e.g.
```java
	public Pattern compileCompleteRegex() {
		Pattern.compile(getCompleteRegex(), Pattern.DOTALL | Pattern.CASE_INSENSITIVE);
	}
```

Reply:
I thought about returning a `Pattern` rather than a string, but thought the string would be more versatile.

I'll try wrapping the entire regex string in a group and adding the dotall and case sensitivity flags.  If it passes the unit tests, I'll keep it in.

Reply:
It looks like adding in the global flags to the regex works - so we'll just keep the string but add the global modifiers as suggested.

Reply:
> I thought about returning a Pattern rather than a string, but thought the string would be more versatile.

Agreed.  My suggestion here was to do both, though moving to inline flags (rather than programmatically set ones) makes that less necessary.